File: packages/claude-code-toolbox/src/agents/runtime/protocols/planThenCode.ts
/**
 * Plan-then-code — the flagship Phase-3 protocol.
 *
 * Flow:
 *   1. PLAN phase: plan agents run round-robin; final message must contain
 *      `<plan>…</plan>`; extracted to `plan.md`.
 *   2. Emit `phase_boundary {needsApproval:true}` and block via
 *      `ctx.awaitApproval(planPath)` — the host resolves this once the user
 *      clicks Approve (optionally with an edited plan).
 *   3. CODE phase: code agents run round-robin, each seeing the approved plan
 *      via --append-system-prompt. Judge (if present on team) runs a final
 *      review turn on the code transcript.
 *
 * If the user rejects, we emit `run_end` with status `aborted` (the reason is
 * surfaced as a `log` event first).
 */
import type { Protocol } from "../runtimeTypes";
import { nowIso } from "../eventTypes";
import {
  describeTranscriptForJudge,
  emitMessage,
  extractPlan,
  findAgentByName,
  makeTotals,
  runOneTurn,
  writePlanArtifact,
} from "./shared";

export const planThenCode: Protocol = async (ctx) => {
  const totals = makeTotals();
  const planAgents = ctx.team.agents
    .map((n) => findAgentByName(ctx.agents, n))
    .filter((a): a is NonNullable<typeof a> => a !== undefined);
  const codeAgents = (ctx.team.codePhaseAgents || [])
    .map((n) => findAgentByName(ctx.agents, n))
    .filter((a): a is NonNullable<typeof a> => a !== undefined);
  if (!planAgents.length) {
    return { status: "error", totals };
  }

  /* === PLAN PHASE === */
  ctx.bus.emit({
    kind: "phase_boundary",
    t: nowIso(),
    runId: ctx.runId,
    from: "none",
    to: "plan",
    needsApproval: false,
  });
  const planTranscript: { agent: string; text: string }[] = [];
  let turn = 0;
  for (const agent of planAgents) {
    if (ctx.signal.aborted) return { status: "aborted", totals };
    turn += 1;
    const prior = planTranscript.length
      ? describeTranscriptForJudge(planTranscript)
      : "(you open the plan discussion)";
    const isLast = turn === planAgents.length;
    const suffix = isLast
      ? `You are the LAST plan agent — your final message MUST wrap the agreed plan in <plan>…</plan> tags. Make the plan executable and specific; reviewers will approve it before any code runs.`
      : `Contribute your angle. Do NOT emit <plan> tags yet — the last agent will do that.`;
    const prompt = [
      `PLAN PHASE, turn ${turn}/${planAgents.length}. Team task:`,
      "",
      ctx.userPrompt,
      "",
      `Prior plan transcript:\n${prior}`,
      "",
      `You are "${agent.name}" (${agent.role}).`,
      suffix,
    ].join("\n");
    const res = await runOneTurn({
      spawn: ctx.spawnAgentTurn,
      agent,
      prompt,
      runId: ctx.runId,
      turn,
      phase: "plan",
      bus: ctx.bus,
      totals,
      cwd: ctx.cwd,
      signal: ctx.signal,
      claudeBin: ctx.claudeBin,
    });
    if (res.errored) return { status: "error", totals };
    planTranscript.push({ agent: agent.name, text: res.text });
  }
  const rawPlan = planTranscript[planTranscript.length - 1].text;
  let planMd = extractPlan(rawPlan);
  const lastAuthor = planTranscript[planTranscript.length - 1].agent;
  const planPath = await writePlanArtifact(ctx.runDir, planMd, ctx.bus, lastAuthor);

  /* === APPROVAL GATE === */
  if (!ctx.awaitApproval) {
    ctx.bus.emit({
      kind: "log",
      t: nowIso(),
      runId: ctx.runId,
      level: "warn",
      message: "No approval handler wired; auto-approving the plan.",
    });
  }
  ctx.bus.emit({
    kind: "phase_boundary",
    t: nowIso(),
    runId: ctx.runId,
    from: "plan",
    to: "code",
    needsApproval: true,
    planPath,
  });
  let approval: { decision: "approve" | "reject"; reason?: string; editedPlan?: string } = {
    decision: "approve",
  };
  if (ctx.awaitApproval) {
    approval = await ctx.awaitApproval(planPath);
  }
  if (approval.decision === "reject" || ctx.signal.aborted) {
    ctx.bus.emit({
      kind: "log",
      t: nowIso(),
      runId: ctx.runId,
      level: "info",
      message: `Plan rejected${approval.reason ? `: ${approval.reason}` : ""}`,
    });
    return { status: "aborted", totals, planArtifactPath: planPath };
  }
  if (approval.editedPlan && approval.editedPlan.trim()) {
    planMd = approval.editedPlan.trim();
    await writePlanArtifact(ctx.runDir, planMd, ctx.bus, `${lastAuthor}+user`);
  }

  /* === CODE PHASE === */
  const phaseAgents = codeAgents.length ? codeAgents : planAgents;
  const codeTranscript: { agent: string; text: string }[] = [];
  for (const agent of phaseAgents) {
    if (ctx.signal.aborted) return { status: "aborted", totals, planArtifactPath: planPath };
    turn += 1;
    const prior = codeTranscript.length
      ? describeTranscriptForJudge(codeTranscript)
      : "(you are the first code-phase agent)";
    const prompt = [
      `CODE PHASE, turn ${turn - planAgents.length}/${phaseAgents.length}. Team task:`,
      "",
      ctx.userPrompt,
      "",
      `APPROVED PLAN (follow exactly — do NOT rewrite):`,
      "",
      planMd,
      "",
      `Prior code transcript:\n${prior}`,
      "",
      `You are "${agent.name}" (${agent.role}). Execute your slice of the plan. Ship small, reviewable edits; announce what you changed.`,
    ].join("\n");
    const res = await runOneTurn({
      spawn: ctx.spawnAgentTurn,
      agent,
      prompt,
      runId: ctx.runId,
      turn,
      phase: "code",
      bus: ctx.bus,
      totals,
      cwd: ctx.cwd,
      signal: ctx.signal,
      claudeBin: ctx.claudeBin,
    });
    if (res.errored) return { status: "error", totals, planArtifactPath: planPath };
    codeTranscript.push({ agent: agent.name, text: res.text });
    const next = phaseAgents[phaseAgents.findIndex((a) => a.name === agent.name) + 1];
    if (next) {
      emitMessage(ctx.bus, agent.name, next.name, res.text);
    }
  }

  /* Optional judge pass — architect or team.judge reviews final code transcript. */
  if (ctx.team.judge) {
    const judge = findAgentByName(ctx.agents, ctx.team.judge);
    if (judge) {
      turn += 1;
      const judgePrompt = [
        `FINAL REVIEW. You are "${judge.name}".`,
        "",
        `Approved plan:`,
        "",
        planMd,
        "",
        `Code-phase transcript:`,
        describeTranscriptForJudge(codeTranscript),
        "",
        `Verify the plan was executed. Reply with APPROVE or REQUEST_CHANGES plus a 1-paragraph explanation.`,
      ].join("\n");
      const jr = await runOneTurn({
        spawn: ctx.spawnAgentTurn,
        agent: judge,
        prompt: judgePrompt,
        runId: ctx.runId,
        turn,
        phase: "code",
        bus: ctx.bus,
        totals,
        cwd: ctx.cwd,
        signal: ctx.signal,
        claudeBin: ctx.claudeBin,
      });
      if (jr.errored) return { status: "error", totals, planArtifactPath: planPath };
    }
  }

  return { status: "completed", totals, planArtifactPath: planPath };
};

File: packages/claude-code-toolbox-intellij/src/main/kotlin/com/amitchorasiya/claude/toolbox/intellij/agents/runtime/protocols/PlanThenCode.kt
package com.amitchorasiya.claude.toolbox.intellij.agents.runtime.protocols

import com.amitchorasiya.claude.toolbox.intellij.agents.runtime.*
import java.io.File

fun planThenCode(ctx: ProtocolContext): ProtocolResult {
    val totals = RunUsage()
    val roster = ctx.agents
    if (roster.isEmpty()) return ProtocolResult("error", totals)
    var turn = 0
    var planArtifactPath: String? = null

    // Phase 1: Plan
    emitMessage(ctx.bus, "system", "all", "Phase 1: Planning")
    ctx.bus.emit(AgentRunEvent.PhaseBoundary(nowIso(), ctx.runId, "none", "plan", false))

    val planTranscript = mutableListOf<Pair<String, String>>()
    for (agent in roster) {
        if (ctx.abortFlag.get()) return ProtocolResult("aborted", totals)
        turn++
        val prevContext = if (planTranscript.isEmpty()) "" else "\n\nPrevious team input:\n${describeTranscriptForJudge(planTranscript)}"
        val prompt = "Plan-then-code — planning phase.\n\nTask: ${ctx.userPrompt}$prevContext\n\nYou are \"${agent.name}\" (${agent.role.name.lowercase()}). Contribute your planning perspective. Wrap your plan in <plan>…</plan> tags."
        val result = spawnAgentTurn(agent, prompt, ctx.runId, turn, "plan", ctx.bus, totals, ctx.cwd, ctx.claudeBin, ctx.abortFlag)
        planTranscript.add(agent.name to result.text)
        if (result.aborted) return ProtocolResult("aborted", totals)
    }

    val lastPlanText = planTranscript.lastOrNull()?.second ?: ""
    val planMd = extractPlan(lastPlanText)
    if (planMd.isNotBlank()) {
        planArtifactPath = writePlanArtifact(ctx.runDir, planMd, ctx.bus, planTranscript.last().first)
    }

    // Phase 2: Approval gate
    if (ctx.awaitApproval != null && planArtifactPath != null) {
        emitMessage(ctx.bus, "system", "user", "Plan ready — awaiting approval")
        ctx.bus.emit(AgentRunEvent.PhaseBoundary(nowIso(), ctx.runId, "plan", "code", true, planArtifactPath))
        val (decision, reason) = ctx.awaitApproval.invoke(planArtifactPath)
        if (decision == "reject") {
            emitMessage(ctx.bus, "user", "system", "Plan rejected${if (reason != null) ": $reason" else ""}")
            return ProtocolResult("aborted", totals, planArtifactPath)
        }
        emitMessage(ctx.bus, "user", "system", "Plan approved")
    } else {
        ctx.bus.emit(AgentRunEvent.PhaseBoundary(nowIso(), ctx.runId, "plan", "code", false))
    }

    // Phase 3: Code
    emitMessage(ctx.bus, "system", "all", "Phase 2: Code execution")
    val codeAgents = if (ctx.team.codePhaseAgents.isNotEmpty()) {
        ctx.team.codePhaseAgents.mapNotNull { findAgentByName(ctx.agents, it) }.ifEmpty { roster }
    } else roster

    val planContent = if (planArtifactPath != null) try { File(planArtifactPath).readText() } catch (_: Exception) { planMd } else planMd
    val codeTranscript = mutableListOf<Pair<String, String>>()
    for (agent in codeAgents) {
        if (ctx.abortFlag.get()) return ProtocolResult("aborted", totals, planArtifactPath)
        turn++
        val prevCode = if (codeTranscript.isEmpty()) "" else "\n\nPrevious work:\n${describeTranscriptForJudge(codeTranscript)}"
        val prompt = "Plan-then-code — code phase.\n\nOriginal task: ${ctx.userPrompt}\n\nApproved plan:\n$planContent$prevCode\n\nYou are \"${agent.name}\". Implement your part of the plan."
        val result = spawnAgentTurn(agent, prompt, ctx.runId, turn, "code", ctx.bus, totals, ctx.cwd, ctx.claudeBin, ctx.abortFlag)
        codeTranscript.add(agent.name to result.text)
        if (result.aborted) return ProtocolResult("aborted", totals, planArtifactPath)
    }

    // Phase 4: Judge review (optional)
    if (ctx.team.judge != null) {
        val judge = findAgentByName(ctx.agents, ctx.team.judge)
        if (judge != null && !ctx.abortFlag.get()) {
            emitMessage(ctx.bus, "system", judge.name, "Final review")
            turn++
            val allWork = describeTranscriptForJudge(codeTranscript)
            val judgePrompt = "You are the reviewer \"${judge.name}\".\n\nOriginal task: ${ctx.userPrompt}\nPlan:\n$planContent\n\nAll code-phase work:\n$allWork\n\nReview: does the implementation match the plan? Flag any gaps."
            spawnAgentTurn(judge, judgePrompt, ctx.runId, turn, "code", ctx.bus, totals, ctx.cwd, ctx.claudeBin, ctx.abortFlag)
        }
    }

    return ProtocolResult("completed", totals, planArtifactPath)
}
